pkg/genesis/vesting_accounts_test.go

```diff
@@ -0,0 +1,498 @@
+package genesis_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/celestiaorg/celestia-app/app"
+	"github.com/celestiaorg/celestia-app/app/encoding"
+	"github.com/celestiaorg/celestia-app/pkg/genesis"
+	"github.com/celestiaorg/celestia-app/test/util/testfactory"
+	"github.com/celestiaorg/celestia-app/test/util/testnode"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	vestingtypes "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmtime "github.com/tendermint/tendermint/types/time"
+)
+
+const (
+	totalAccountsPerType = 300
+	initBalanceForGasFee = 10
+	vestingAmount        = testfactory.BaseAccountDefaultBalance
+	vestingDelayPerTx    = 10 // this is a safe time (in seconds) to wait for a tx to be executed while the vesting period is not over yet
+)
+
+type accountDispenser struct {
+	names   []string
+	counter int
+}
+
+type accountType int
+
+const (
+	RegularAccountType accountType = iota + 1
+	DelayedVestingAccountType
+	PeriodicVestingAccountType
+	ContinuousVestingAccountType
+)
+
+func (t accountType) String() string {
+	switch t {
+	case RegularAccountType:
+		return "regular"
+	case DelayedVestingAccountType:
+		return "delayed vesting"
+	case PeriodicVestingAccountType:
+		return "periodic vesting"
+	case ContinuousVestingAccountType:
+		return "continuous vesting"
+	default:
+		return "unknown"
+	}
+}
+
+type VestingModuleTestSuite struct {
+	suite.Suite
+
+	// accounts is a map from accountType to accountDispenser
+	accounts    sync.Map
+	accountsMut sync.Mutex
+
+	kr   keyring.Keyring
+	cctx testnode.Context
+	ecfg encoding.Config
+}
+
+func TestVestingModule(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping vesting module test suite in short mode.")
+	}
+	suite.Run(t, new(VestingModuleTestSuite))
+}
+
+func (s *VestingModuleTestSuite) SetupSuite() {
+	t := s.T()
+	t.Log("setting up vesting module test suite")
+
+	s.kr = testfactory.GenerateKeyring()
+
+	genOpts := []testnode.GenesisOption{}
+	genOpts = append(genOpts, s.initRegularAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initDelayedVestingAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initPeriodicVestingAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initContinuousVestingAccounts(totalAccountsPerType))
```

Comment:
as we've discussed previously, it looks like a lot of the logic here is actually things that we need in the generic code to add things to the genesis. If that's the case, then I don't think they should be specific to this test.

more specifically, these functions are both calling the function to create the vesting account, and then adding it to the genesis. We want a generic function that does this exact thing, then we test it.

does this make sense? if not I'm happy to sync on this to get more specific if needed :slightly_smiling_face:

Reply:
well right now it is pretty much like this. It just does it from another perspective.

We have functions like this one which is generic and creates a generic genesis account with its balance with the given input for us.
```go
func NewGenesisDelayedVestingAccount(
	address string,
	vestingBalance,
	initUnlockedCoins sdk.Coins,
	endTime time.Time,
) (account authtypes.GenesisAccount, balance banktypes.Balance, err error) {
...
}
```

and then we just need to call the following function to add it to genesis:

```go
func AddGenesisAccountsWithBalancesToGenesisState(
	encCfg encoding.Config,
	gs map[string]json.RawMessage,
	gAccounts []authtypes.GenesisAccount,
	balances []banktypes.Balance,
) (map[string]json.RawMessage, error)
```

So does it do the same thing or I did not get your point properly?

Reply:
exactly, we're almost doing it, but we just need to go the extra step to do it

literally just combining these functions into a different function that accepts a genesis, list of accout to create, the returns a genesis back